File: web/src/core/ocr/OCRProvider.ts
export interface OCRRecognizeOptions {
  /** Restrict recognition to these characters, e.g. digits only. */
  whitelist?: string;
}

export interface OCRResult {
  text: string;
  /** 0-100 engine confidence. */
  confidence: number;
}

/**
 * Abstraction over the local OCR engine, so the Tesseract.js-based
 * implementation can be swapped for a different local engine later
 * without touching RegionExtractor/ReviewTable/validation code. No
 * provider may call a cloud/network OCR service - see TesseractOCRProvider
 * for how offline operation is enforced (vendored worker/core/lang files).
 */
export interface OCRProvider {
  initialize(): Promise<void>;
  recognize(image: HTMLCanvasElement, options?: OCRRecognizeOptions): Promise<OCRResult>;
  terminate(): Promise<void>;
}

File: app/src/main/java/com/productionboard/scanner/ocr/OCRProvider.kt
package com.productionboard.scanner.ocr

import android.graphics.Bitmap

data class OCRResult(val text: String, val confidence: Float)

/**
 * Abstraction over the local OCR engine, so the Tesseract-based
 * implementation ([TesseractOCRProvider]) can be swapped for a different
 * fully-offline engine later without touching RegionExtractor/
 * ReviewScreen/OCRValidator. No implementation may call a network/cloud
 * OCR service.
 */
interface OCRProvider {
    suspend fun initialize()
    suspend fun recognize(bitmap: Bitmap, whitelist: String? = null): OCRResult
    fun close()
}
